cas-server-core-tickets/src/main/java/org/jasig/cas/ticket/OAuthCodeImpl.java

```diff
@@ -0,0 +1,118 @@
+package org.jasig.cas.ticket;
+
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.jasig.cas.authentication.Authentication;
+import org.jasig.cas.authentication.principal.Service;
+import org.jasig.cas.ticket.proxy.ProxyGrantingTicket;
+import org.springframework.util.Assert;
+
+import javax.persistence.Column;
+import javax.persistence.DiscriminatorColumn;
+import javax.persistence.DiscriminatorValue;
+import javax.persistence.Entity;
+import javax.persistence.Lob;
+import javax.persistence.Table;
+import javax.validation.constraints.NotNull;
+
+/**
+ * An OAuth code implementation.
+ *
+ * @author Jerome Leleu
+ * @since 4.3.0
+ */
+@Entity
+@Table(name="OAUTH_TOKENS")
+@DiscriminatorColumn(name="TYPE")
+@DiscriminatorValue(OAuthCode.PREFIX)
+public class OAuthCodeImpl extends AbstractTicket implements OAuthCode {
```

Comment:
So one solution would be not hardcode delegators into the registry, but move them out into a component that can be extended via OAuth code, so it can in turn build up its own delegators and pass them to the core ticket registry.